### PR TITLE
Adding driver for ADG731 (Analog Devices 32-channel analog multiplexer) 

### PR DESCRIPTION
**Pull request by andrasfuchs**
This is the addition of the driver for ADG731 multiplexer IC with a sample demonstrating channel switching from 0 to 31.

### CODE REVIEW

File: src/devices/Adg731/Adg731.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Device.Spi;
+
+namespace Iot.Device.Adg731
```
Comment (krwq) [inline src/devices/Adg731/Adg731.cs:4]: namespace should be `Iot.Device.Adc` - this will make it easier to use since class name and namespace have different name

File: src/devices/Adg731/Adg731.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
+﻿using System;
```
Comment (krwq) [inline src/devices/Adg731/Adg731.cs:1]: copyright header missing

File: src/devices/Adg731/Adg731.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            }
+
+            set
+            {
+                _isEnabled = value;
+                StateChanged();
+            }
+        }
+
+        private bool _isSelected;
```
Comment (krwq) [inline src/devices/Adg731/Adg731.cs:37]: fields should be declared first in the class and be together

File: src/devices/Adg731/Adg731.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                throw new Exception($"SPI device must have a lower clock frequency, because ADG731's maximum operating SPI frequncy is {_metadata.MaximumSpiFrequency} Hz.");
+            }
+
+            _spiDevice = spiDevice;
+            StateChanged();
+        }
+
+        private void StateChanged()
+        {
+            byte[] data = new byte[1];
```
Comment (krwq) [inline src/devices/Adg731/Adg731.cs:106]: please use:

```csharp
Span<byte> data = stackalloc byte[1];
```

instead

File: src/devices/Adg731/Adg731.csproj [hunk 5]
```diff
@@ -0,0 +1,33 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (krwq) [inline src/devices/Adg731/Adg731.csproj:1]: please make sure content of csproj is as similar to other csproj-es we have

File: src/devices/Adg731/BitMask.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿namespace Iot.Device.Adg731
+{
+    /// <summary>
+    /// ADG731 Bit-masks
+    /// </summary>
+    public enum BitMask : byte
```
Comment (krwq) [inline src/devices/Adg731/BitMask.cs:6]: this should not be public

File: src/devices/Adg731/BitMask.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        /// <summary>
+        /// Chip Enable
+        /// </summary>
+        EN = 0b1000_0000,
```
Comment (krwq) [inline src/devices/Adg731/BitMask.cs:11]: please expand abbreviations of names here (since you only got 3 items)

File: src/devices/Adg731/BitMask.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿namespace Iot.Device.Adg731
```
Comment (krwq) [inline src/devices/Adg731/BitMask.cs:1]: copyright header

File: src/devices/Adg731/Metadata/Adg731Metadata.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
+﻿using System.Device.Spi;
```
Comment (krwq) [inline src/devices/Adg731/Metadata/Adg731Metadata.cs:1]: copyright header missing

File: src/devices/Adg731/Metadata/Adg731Metadata.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Iot.Device.Adg731
+{
+    /// <summary>
+    /// Metadata class for ADG731
+    /// </summary>
+    public class Adg731Metadata : IDeviceMetadata, ISpiDeviceMetadata, IMultiplexerDeviceMetadata
```
Comment (krwq) [inline src/devices/Adg731/Metadata/Adg731Metadata.cs:8]: please remove the interfaces and create an issue instead for them so that we either add it to all devices or everywhere or nowhere - regardless we should be consistent with what the rest of devices does

File: src/devices/Adg731/Metadata/Adg731Metadata.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        public string Manufacturer => "Analog Devices";
+
+        /// <inheritdoc/>
+        public string Product => "ADG731";
+
+        /// <inheritdoc/>
+        public string ProductCategory => "Analog Multiplexer";
+
+        /// <inheritdoc/>
+        public string ProductDescription => "32-Channel, Serially Controlled 4 Ohm 1.8 V to 5.5 V, +/- 2.5 V, Analog Multiplexer";
```
Comment (krwq) [inline src/devices/Adg731/Metadata/Adg731Metadata.cs:20]: IMO:
- datasheet link should be in comment and README.md
- ValidSpiModes should be an IEnumerable and be directly on the ADC (as static method: i.e. GetValidSpiModes which can return cached array or something) - reason for IEnumerable is because SpiMode is not `[Flags]` enum
-

File: src/devices/Adg731/samples/Adg731.Samples.BasicSample.csproj [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    </ProjectReference>
+    <ProjectReference Condition="$(Configuration.Contains('Linux'))" Include="$(MainLibraryPath)System.Device.Gpio.csproj">
+      <AdditionalProperties>$(AdditionalProperties);RuntimeIdentifier=linux</AdditionalProperties>
+    </ProjectReference>
+    <!--Cases used when running build script in other cases. Just pick one -->
+    <ProjectReference Condition="'$(Configuration)' == 'Debug' or '$(Configuration)' == 'Release'" Include="$(MainLibraryPath)System.Device.Gpio.csproj">
+      <AdditionalProperties>$(AdditionalProperties);RuntimeIdentifier=linux</AdditionalProperties>
+    </ProjectReference>
+    <ProjectReference Include="..\Adg731.csproj" />
+    <PackageReference Include="Iot.Device.Bindings" Version="1.0.0" />
```
Comment (krwq) [inline src/devices/Adg731/samples/Adg731.Samples.BasicSample.csproj:22]: sample should follow naming conventions (including csproj name and content) as in other samples - this will currently have chicken and the egg problem: Iot.Device.Bindings doesn't have your library yet because we haven't shipped package yet and we won't until the build passes - this should do what every other project does: use project to project reference - ideally copy&paste content of other sample and modify as little as possible

File: src/devices/Adg731/samples/Adg731.Samples.BasicSample.csproj [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    <PackageReference Include="NetCoreServer" Version="3.0.7" />
```
Comment (krwq) [inline src/devices/Adg731/samples/Adg731.Samples.BasicSample.csproj:23]: please try not using external packages unless it's absolutely required by device. In case of ADC it is not.

File: src/devices/Adg731/samples/DeployToPi.ps1 [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+function OutputDataReceived {
+	param($e)
+
+	Write-Host $e.Data
+}
+
+# Set up session options
+$sessionOptions = New-Object WinSCP.SessionOptions -Property @{
```
Comment (krwq) [inline src/devices/Adg731/samples/DeployToPi.ps1:28]: this file probably shouldn't be here...

File: src/devices/Adg731/samples/Program.cs [hunk 15]
```diff
@@ -0,0 +1,67 @@
+﻿using System;
```
Comment (krwq) [inline src/devices/Adg731/samples/Program.cs:1]: copyright header

File: src/devices/Adg731/samples/Program.cs [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+using System.Device.Spi;
+using System.Diagnostics;
+using System.Threading;
+using Iot.Device.Adg731;
+
+namespace ADG731Sample
+{
+    internal class Program
+    {
+        private static Adg731 s_adg731;
```
Comment (krwq) [inline src/devices/Adg731/samples/Program.cs:11]: sample can probably be written so it doesn't have static fields

File: src/devices/Adg731/samples/Program.cs [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+            SpiDevice adg731SpiDevice = SpiDevice.Create(settings);
+
+            Console.WriteLine($"Connecting to ADG731 using SPI {adg731SpiDevice.ConnectionSettings.Mode.ToString()} at {adg731SpiDevice.ConnectionSettings.ClockFrequency / 1000.0:N1} kHz...");
+
+            s_adg731 = new Adg731(adg731SpiDevice);
+            s_adg731.IsEnabled = true;
+            s_adg731.IsSelected = true;
+
+            int loopcounter = 0;
+            while (true)
```
Comment (krwq) [inline src/devices/Adg731/samples/Program.cs:34]: I suggest to do regular for loop and skip condition: `for (int i = 0; ; i++)`

File: src/devices/Adg731/samples/Program.cs [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                s_adg731.ActiveChannel = loopcounter % ((IMultiplexerDeviceMetadata)Adg731.GetDeviceMetadata()).MultiplexerChannelCount;
+                Console.WriteLine($"Channel changed to {s_adg731.ActiveChannel}");
+                loopcounter++;
+                Thread.Sleep(500);
+            }
+        }
+
+        private static void WaitForDebugger()
```
Comment (krwq) [inline src/devices/Adg731/samples/Program.cs:43]: shouldn't be part of the sample (btw you can search for Scott Hanselman post about debugging on Pi - link should be somewhere here in the repo)

File: src/devices/Adg731/samples/README.md [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+* [Pmod HAT Adapter](https://reference.digilentinc.com/reference/add-ons/pmod-hat/reference-manual) (Raspberry Pi Pmod HAT)
+* [Extech MN47 multimeter](http://www.extech.com/products/MN47) (for verifying switch completion)
+
+## Test unit and signal input
+![Raspberry Pi 3 Model B+ + Digilent Pmod HAT + BBD Mars-64 front side](Raspberry_3Bp_PmodHAT_BBDMars64_front.jpg)
+
+![Raspberry Pi 3 Model B+ + Digilent Pmod HAT + BBD Mars-64 back side](Raspberry_3Bp_PmodHAT_BBDMars64_back.jpg)
+
+
+## Installation on Raspberry Pi
```
Comment (krwq) [inline src/devices/Adg731/samples/README.md:19]: you can go ahead and create separate PR for deploying to PI instructions which can be used globally but please let's not add them per sample.

File: src/devices/Adg731/samples/README.md [hunk 20]
```diff
@@ -0,0 +1,102 @@
+# ADG731 - Samples
+
+## Required hardware
+* [Analog Devices ADG731](https://www.analog.com/media/en/technical-documentation/data-sheets/ADG725_731.pdf)
+* Any supported IoT device with an SPI bus
```
Comment (krwq) [inline src/devices/Adg731/samples/README.md:5]: actually that's not fully true, you it will probably work on any device with GPIO as well since we support software SPI, also will work with any PC which has FT4222 USB dongle which exposes SPI through one of our bindings. I'd skip the required board section per device since this can easily get out of date in the future

File: src/devices/Adg731/samples/README.md [hunk 21]
```diff
@@ -0,0 +1,102 @@
+# ADG731 - Samples
```
Comment (krwq) [inline src/devices/Adg731/samples/README.md:1]: I'd limit this file to purely code - how to use it and go through interesting scenarios and not debugging and setting up PI. For those instructions please add them to documentation folder somewhere in separate PR

File: src/devices/Adg731/samples/README.md [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	Console.WriteLine();
+}
+```
+
+## Result
+You can see that the device is switched between the channels from channel 0 to 31.
+
+![Results in a PowerShell window](PowerShell_RaspberryPi_Output.png)
```
Comment (krwq) [inline src/devices/Adg731/samples/README.md:102]: can't put comment there: please remove checked in DLL. Edit README.md on your binding (not sample/README.md) to mention where to get WinSCP and why it's needed. Assuming it's for deployment you should put that in separate PR with instructions how to deploy to PI - binaries in the repo are very frowned upon in the open source community.

File: src/devices/Adg731/Adg731.cs (hunk 1 above)
Comment (krwq) [reply src/devices/Adg731/Adg731.cs:4]: Actually reading the code further and docs a bit this should be int `.Multiplexer` namespace. Also device should expose way of selecting input channel with API which should be crystal clear i.e.:
`int SelectedChannel { get; set; }`

**Comment by krwq** [comment]
@andrasfuchs do you think you'll have time to address this?

**Comment by andrasfuchs** [comment]
> @andrasfuchs do you think you'll have time to address feedback here?

Sorry, I don't think I'll be able to work on this in the near future. 
Feel free to make the suggested modifications or discard the pull request.

**Comment by joperezr** [comment]
[Triage] PR seems abandoned and author won't have time to finish it. We will create an issue for this binding and attach the progress made on this PR so if anybody is interested in the future they can build upon the progress that was made here. We will close this PR next week if there is no activity by then.
